### DISCUSSION TITLE
Make it easier to view API Swagger Definitions

### DISCUSSION THREAD

**Discussion by mffap** [zitadel/zitadel#2645] [Ideas]
We publish [Swagger Definitions](https://docs.zitadel.ch/docs/apis/introduction#swagger-documentation) on our docs page. But to view them you need an external editor / SwaggerUI.

Naturally https://editor.swagger.io/ comes to mind, which offers a method to import from url.

![image](https://user-images.githubusercontent.com/1366906/140931138-baf1ad6a-a6b4-4198-89dc-b09202275728.png)

However this is being blocked by CORS policy

![image](https://user-images.githubusercontent.com/1366906/140931191-d2bf225f-cb0a-4d50-b752-b17c216412ab.png)

A workaround is to download the file and import the file to the editor. 

For me this feels cumbersome. Would it be feasible to make the import work? Or could we show the definitions right from docs.zitadel.ch?

**Comment by fforootd** [comment]
Well, we can either:

- Change the CORS policy for that specific endpoint (@livio-a IMO no big deal)
- Host an openapi gui (I am no fan of this)
- Publicly upload the API Spec after a release to e.g. [Postman's Public API Network ](https://www.postman.com/explore)

**Reply by livio-a** [reply]
For the moment, I'd suggest opening CORS on that endpoint to (origin: *) and change / add a link in the docs, which open Swagger Editor with the corresponding API, e.g.: https://editor.swagger.io/?url=https://raw.githubusercontent.com/OAI/OpenAPI-Specification/master/examples/v3.0/petstore.yaml

We can always publish the API later on. And I'm no fan of hosting an openapi either.

**Reply by fforootd** [reply]
Yes, so let's first open up the cors settings

**Reply by livio-a** [reply]
Opened an new issue:
-  https://github.com/caos/zitadel/issues/2655

**Comment by Frontrider** [comment]
The inability to just browse/simply download that api (and then try to understand it while actually using it) is gonna be the final reason why I'll stop trying to use this.

**Comment by corepay** [comment]
I'm confused here.  Is the openapi spec available for Zitadel endpoints?  @mffap link above is dead and @Frontrider 's ungrateful threat tells me the spec is not available.

I would like to get it so I can try out different endpoints without having to rebuild each request in Insomnia.  The zitadel api docs have a SEND REQUEST feature but no way I can find to change the base url so they are useless.

BTW - highly recommend @mffap or whoever is in charge of the API docs take a serious look at https://scalar.com/

**Comment by Frontrider** [comment]
> I'm confused here. Is the openapi spec available for Zitadel endpoints? @mffap link above is dead and @Frontrider 's ungrateful threat tells me the spec is not available.
> 
> I would like to get it so I can try out different endpoints without having to rebuild each request in Insomnia. The zitadel api docs have a SEND REQUEST feature but no way I can find to change the base url so they are useless.
> 
> BTW - highly recommend @mffap or whoever is in charge of the API docs take a serious look at https://scalar.com/

There is a spec that is probably also either old, or was never updated to a newer spec (which is not a problem on it's own), but it is presented in a state that I could not turn into anything useful without scripting more on top of it.
https://zitadel.cloud/openapi/v2/swagger/v2/zitadel/zitadel/

The way it is split and provided makes it more work then its worth. Especially bad, as this would be the single most important document the api it can have, even more important than any provided sdk. (as it IS the SDK for any platform)

Also, that link is buried deep in the docs, and is not easy to get to. I think there is also a second link, but can't find it anymore. That was even worse to try to read.

**Reply by fforootd** [reply]
Thanks for sharing your thoughts. The main reason why it is structure in a unconventional way for openapi is because we generate the openapi spec out of the [gRPC/proto definition](https://github.com/zitadel/zitadel/tree/main/proto/zitadel) which can build clients for many platform with the help of buf.

I understand though that this leaves some room open for improvements which I will look into if there is an easy solution.

@corepay can Insomnia not import a folder of files in openapi and generate the requests for you?
You could import or reflect the gRPC api https://docs.insomnia.rest/insomnia/grpc#use-grpc-server-reflection as alternative

**Comment by corepay** [comment]
Hi Florian, thanks for reply.

I am not interested in gRpc - haven't used it and am well qualified in REST.  My whole platform is riding a restful openaapi spec and I don't want to mix parts with different params and protocols - also dont really want to mix code my backend with grPC right now as wonderful as I hear it is.  gRPC 'can' build clients for 'many' clients 'with' the help of buf as I understand...but REST is universal across all clients afaik and client generators are mature in openapi tooling.  I'm sure there are a million reasons why Zitadel emphasizes gRPC but for me - an 95%er non-enterprise tire-kicker gRPC is too complex or at least yet another tech I need to stuff in my overloaded and overworked brain and overkill for the auth portion of my app (my .02) - I am finding success reasoning the PAI docs and workign with the cURL as a copy/paste in insomnia to generate the REST body and query params at least.

Like the dreaded typescript I might grudingly be forced to adopt grpc soomeday but for now...I'm sticking with REST!

Thank you for such a well-thought out feature rich authentication tool.  I think you would benefit a lot with a fresh look at your docs, I see you have pretty much everything covered but I am finding myself with a dozen tabs open clicking in and out to find little pieces of related data for even simple tasks like introspection.   I wrote a little diddy a few days ago about link hell.  I think you got too much information (not a bad thing) that is not well organized (bad thing).  

My suggestion is to focus like a hawk on the API and provide a simple, organized REST openapi spec to attract more interest and reduce these long winded question/responses.  The ONLY reason Stripe got so popular so fast was they made payment integration simple for developers.  REST is universal, I am in the fintech and have integrated literally 40 or 50 payment gateways  and related products and wrote APIs for a dozen or more clients ALL of them have 1st class support for REST.  Gotta be honest as a guy on the street I am really struggling here, this has been be far the hardest integration and I'm still not sure how far I can or should go with a v2 "beta" and v3 "preview" all mixed in both your docs and API spec.  

Writing openapi spec is not hard - don't rely on a lib to do get it done - these things are a work of art.  If you don't have an expert or time to do it I might could help.

I hope you take this constructively there is no mal intent.   I just want to see you guys fly...saw ya'll get started probably close to 3 years ago by now, was intrigues then and found some time and a reason to give it a run - the tenancy is pretty unique and almost fits exactly what I need.

**Reply by fforootd** [reply]
No insult take, we actually like if people clearly state what they think would work for them 🤩 

So let us do it like that, I will take some time later this week and see if we can improve the openapi piece (in one way or the other).

**Comment by sandro-yiuuio** [comment]
I know that for some gRPC is the holy grail because of different reasons. But in all honesty, as soon as you get out of your own project the lingua franca of integration is REST and not gRPC.

In all honesty... I LOVE Zitadel from a conceptual perspective in CAPITAL letters. However, we found that the API is maybe one of the worst ones we have ever used. It's segmented into three different versions with overlaps, functional gaps, different approaches and concepts and counterintuitive ways of doing things (like putting the user's account id info into the header if doing things in behalf of another account). It feels like an unloved appendix to the system. It's a bit of a shame that an outstanding system like Zitadel lacks so much under the hood when it comes to the API. And let's be honest... it's not a rocket science piece of thing but mostly data in - data out. How about bringing this baby to the next level? Implement a real and easy and comprehensive REST interface to Zitadel in a consistent way for all relevant objects along with making it available through a proper OpenAPI portal? THAT's the state of the art of integration. And Zitadel would be an excellent candidate for integration scenarios if the REST API would be as high quality as the rest.

**Comment by lakmalniranga** [comment]
First of all, ZITADEL is one of the easiest auth servers to get started with. However, as soon as you need to use the API, it becomes painful. I initially used the Java SDK, but it was too verbose, so I switched to the REST API. Unfortunately, the lack of an OpenAPI specification makes things very difficult, especially when trying to build  with LLMs.

**Comment by Frontrider** [comment]
On my side, it was easier to drop this compared to attempting to figure it out.
